### PR TITLE
[prototype] Expert Parallel

### PR DESCRIPTION
**Pull request by tianyu-l**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #714

The expert-choice MoE implementation is mostly from torchtune: https://github.com/pytorch/torchtune/pull/1902
The PR requires (ad hoc) changes to pytorch: https://github.com/pytorch/pytorch/pull/141937

Issue tracking:
- [ ] [dp2ep] how to apply FSDP only to the non-MoE modules? (for now need to comment out `fully_shard`)
- [ ] [dp2ep] `shard_dim_alltoall` not robust (especially during backward with more than 1D)
- [ ] [tp2ep] backward efficiency may not optimized (e.g. right now `aten.scatter.src` only supports [replicate](https://github.com/pytorch/pytorch/blob/main/torch/distributed/tensor/_ops/_tensor_ops.py#L368) sharding prop)
- [ ] [tp2ep] using DTensor (e.g. in "tp2ep"), the backward `aten.scatter.src` requires `_allow_implicit_replication` (maybe because in backward some tensor is not generated as DTensor)
- [ ] some other issues tracked in https://github.com/pytorch/pytorch/pull/141937
- [ ] `torch.compile` fails on `torch.topk`

Haven't worked on
- softmax scoring instead of sigmoid (can be done similarly, would incur extra communications)
- part of DP (e.g. CP) to EP

Not considering
- shared expert overlapping
- token-choice MoE

### CODE REVIEW

File: torchtitan/models/llama/moe_layers.py [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+                ``(num_experts*tokens_per_expert, dim)``.
+            token_indices (torch.Tensor): token indices for routed_input. Shape ``(num_experts*tokens_per_expert,)``.
+        """
+        # scores shape (num_experts, bs*slen)
+        scores = self.gate(x).transpose(0, 1)
+        # By default, we perform sigmoid and softmax in float32 to avoid loss explosion.
+        if self.use_sigmoid:
+            scores = torch.sigmoid(scores.to(torch.float32)).to(x.dtype)
+        else:
+            scores = F.softmax(scores.to(torch.float32), dim=0).to(x.dtype)
```
Comment (lessw2020) [inline torchtitan/models/llama/moe_layers.py:132]: one comment here - it's not clear imo if the optimal order is softmax and then topk, or topk and then softmax.  
It does not change the routing, but it changes the weights. 
In my nanogpt MoE it has the option to toggle the ordering.

**Comment by tianyu-l** [comment]
somehow I lost access to this ghstack, moving to https://github.com/pytorch/torchtitan/pull/725 instead
